File: JavaScript/sort/BrickSort.js
/*

A sorting algorithm which is famously known as odd-even sort ( or Parity sort or Brick sort ) is a modification of bubble sort.
Algorithm -> Divide into the two phases - 
             1. Odd Phase ( Perform bubble sort on odd indexed elements )
             2. Even Phase ( Perform bubble sort on even indexed elements )

Time Complexity -> O(N^2) where, N is the number of elements in the input array

*/

// Prompt node package for taking user input
const prompt = require("prompt-sync")({ sigint: true });

// perform bubblesort operation w.r.t to the passed indexes
function bubbleSort(array, start, end) {
  let isSorted = true;
  for (let i = start; i <= end; i += 2) {
    if (array[i] > array[i + 1]) {
      [array[i], array[i + 1]] = [array[i + 1], array[i]];
      isSorted = false;
    }
  }

  return isSorted;
}

// perform bricksort
function brickSort(array) {
  let isSorted = false;

  while (!isSorted) {
    isSorted = true;

    isSorted = bubbleSort(array, 1, array.length - 2);
    isSorted = bubbleSort(array, 0, array.length - 2);
  }

  return array;
}

/* Workflow of user input */

// Take array length as input
let arrayLength = +prompt("Enter array length - ");

// Check whether the entered value is number or not
if (isNaN(arrayLength)) return console.log("Only numbers are allowed");

// Globally declared array
let array = [];

// Take array items
for (let i = 1; i <= arrayLength; i++) {
  array.push(+prompt(`Enter ${i} element - `));

  // Check whether the entered value is number or not
  if (array.includes(NaN)) return console.log("Only numbers are allowed");
}

console.log("Your array - ", array);

// Call the algorithm
brickSort(array);
console.log("Sorted array", array);

// Sample I/O

/*

> node BrickSort

Enter array length - 7
Enter 1 element - 56
Enter 2 element - 89
Enter 3 element - 40
Enter 4 element - 60
Enter 5 element - 21
Enter 6 element - 90
Enter 7 element - 10
Your array -  [
  56, 89, 40, 60,
  21, 90, 10
]
Sorted array [
  10, 21, 40, 56,
  60, 89, 90
]

*/

File: C-Plus-Plus/sort/Brick_Sort.cpp
/*
In our program in each iteration we first do bubble sort on odd indexed elements and then a bubble sort on the even indexed elements.
Will we get a sorted result if we first perform a bubble sort on even indexed element first and then on the odd indexed element
*/

#include<bits/stdc++.h>
using namespace std;

// A function to sort the algorithm using Odd Even sort
void oddEvenSort(int arr[], int n)
{
    // Initially array is unsorted
    bool isSorted = false; 
 
    while (!isSorted)
    {
        isSorted = true;
 
        //Bubble sort on odd indexed element
        for (int i=1; i<=n-2; i=i+2)
        {
            if (arr[i] > arr[i+1])
             {
                //swap arr[i] and arr[i+1]
                int temp=arr[i];
                arr[i]=arr[i+1];
                arr[i+1]=temp;
                
                isSorted = false;
              }
        }
 
        // Bubble sort on even indexed element
        for (int i=0; i<=n-2; i=i+2)
        {
            if (arr[i] > arr[i+1])
            {
                //swap arr[i] and arr[i+1]
                int temp=arr[i];
                arr[i]=arr[i+1];
                arr[i+1]=temp;
                isSorted = false;
            }
        }
    }
 
    return;
}

signed main()
{
    //Taking Input
    int n; cin>>n;
    int arr[n];
    for(int i=0;i<n;i++){
        cin>>arr[i];
    }

    //Applying Sort
    //calling sort function
    oddEvenSort(arr, n);
    cout<<"Sorted Array is :-";
    for (int i = 0; i < n; i++){ 
        cout<<arr[i]<<" ";
    }
    return 0;
}
/* Sample Input
11
35 23 12 17 12 72 31 46 180 88 54
*/
/* Sample Output
Sorted Array is:- 12 12 17 23 31 35 46 54 72 88 180
*/

// Time Complexity -O(n^2)
//Space Complexity -O(1)
